pkg/ui/server.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ui
+
+import (
+	"context"
+	"html/template"
+	"log"
+	"net/http"
+	"strings"
+	"time"
+)
+
+// Pair represents a key value pair used by the select HTML element.
+type Pair struct {
+	Key  string
+	Text string
+}
+
+// Content defines the displayable parts of the token retrieval form.
+type Content struct {
+	UserLabel     string
+	CategoryLabel string
+	ReasonLabel   string
+	TTLLabel      string
+	Categories    []Pair
+	TTLs          []Pair
+}
+
+// FormDetails represents all the input and content used for the token retrievlal form.
+type FormDetails struct {
+	WindowName string
+	Origin     string
+	PageTitle  string
+	Content    Content
+	Category   string
+	Reason     string
+	TTL        string
+	Errors     map[string]string
+}
+
+// SuccessDetails represents the data used for the success page and the postMessage response to the client.
+type SuccessDetails struct {
+	Token        string
+	TargetOrigin string
+	WindowName   string
+}
+
+var (
+	categories []string
+	ttls       []string
+)
+
+// RunServer initializes a server on port 9091 and registers a handler for the /popup route.
+func RunServer(ctx context.Context) {
+	categories = []string{"explanation", "breakglass"}
+	ttls = []string{"15", "30", "60", "120", "240"}
+
+	router := http.NewServeMux()
+	server := &http.Server{
+		Addr:         ":9091",
+		Handler:      router,
+		ReadTimeout:  5 * time.Second,
+		WriteTimeout: 1 * time.Second,
+		IdleTimeout:  15 * time.Second,
+	}
+
+	fs := http.FileServer(http.Dir("./assets/static"))
+	router.Handle("/assets/static/", http.StripPrefix("/assets/static/", fs))
+	router.HandleFunc("/popup", popup)
+	log.Fatal(server.ListenAndServe())
+}
+
+func popup(w http.ResponseWriter, r *http.Request) {
```

Comment:
This should be `HandlePopup` and return a `http.Handler`

Reply:
In my upcoming commit I mimic the metrics aggregator so this will look slightly different. Let me know if the latest commit still requires a capitalized function name.

Reply:
Hey sorry I should have been clearer. It's not the capitalization (uppercase functions are public, lowercase are private). It's about the return type. We want to return a http.Handler instead of the function itself being an http.Handler.

Reply:
I spoke with Shan, he will mimic the router pattern in metrics-aggregator.